### PR TITLE
Blog App: validations, model specs, and n+1 problems

### PR DESCRIPTION
**Pull request by Kalunge**
## Blog App: validations, model specs, and n+1 problems
- Add the following validations:
  - For the User model:
    - Name must not be blank.
    - PostsCounter must be an integer greater than or equal to zero.
  - For the Post model:
    - Title must not be blank.
    - Title must not exceed 250 characters.
    - CommentsCounter must be an integer greater than or equal to zero.
    - LikesCounter must be an integer greater than or equal to zero.
- Add unit specs for all of your models' methods and validations.
- Add flash messages in the create actions in all your controllers.
- Make sure that the N+1 problem is solved when fetching all posts and their comments for a user. This is the corresponding wireframe view:

### CODE REVIEW

File: spec/models/post_spec.rb [hunk 1]
```diff
@@ -1,5 +1,20 @@
...
+    post = Post.new(user: nil)
+    expect(post).to_not be_valid
+  end
+
+  it 'Title must not be too long' do
+    user = User.create(Name: 'John Doe', Bio: 'Software developer')
+    title = 'hello' * 100
+    post = Post.create(Title: title, user: user)
+    expect(post).to_not be_valid
+  end
```
Comment (Dipeshtwis) [inline spec/models/post_spec.rb]: - [ ] I would recommend adding the test for the positive scenarios as well. Ex:


```ruby
describe 'post title validations' do
  it { should validate_presence_of(:title) }
end
```

File: spec/models/user_spec.rb [hunk 2]
```diff
@@ -1,5 +1,22 @@
...
+  end
+
+  it 'is not valid if photo absent' do
+    @user.photo = nil
+    expect(@user).to_not be_valid
+  end
+
+  it 'is not valid without bIO' do
+    @user.Bio = nil
+    expect(@user).to_not be_valid
```
Comment (Dipeshtwis) [inline spec/models/user_spec.rb]: - [ ] Here as well. I would recommend adding the test for the positive scenarios as well. Ex:


```ruby
describe 'user name validations' do
  it { should validate_presence_of(:name) }
end
```

File: README.md [hunk 3]
```diff
@@ -25,9 +26,8 @@ To get a local copy up and running follow these simple example steps.
 - Clone or download this repo on your machine
   
   ```git clone git@github.com:Kalunge/Blog-app.git```
-- cd into project directory![145261309-cf2fa744-ddfb-4c37-bf6b-5a5ebf6ea28d](https://user-images.githubusercontent.com/50773868/145261402-cc401dbf-f42f-4b09-b291-5605678952ee.png)
-
-  
+- cd into project directory
+
```
Comment (Dipeshtwis) [inline README.md:30]: - [ ] The command `rake db:reset` gives me the error. The seed data has some issue, giving the not null violation error. Please fix the seed data.

Screenshot from 2021-12-09 19-56-48

**Review by Dipeshtwis — CHANGES_REQUESTED** [review]
Hi @Kalunge, 

:+1: This is good to make it this far, your project is really nice, congratulations on the good work you have done so far.
:+1: All tests are passing.

Good job so far!
There are some issues that you still need to work on to go to the next project but you are almost there!


 ## Required Changes ♻️

**_Check the comments under the review_**

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

- N.A.

Cheers and Happy coding!👏👏👏
Feel free to leave any questions or comments in the PR thread if something is not 100% clear.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

**_Dipesh Kumar_**


------

_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project(check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: spec/models/user_spec.rb (hunk 2 above)
Comment (henatan99) [reply spec/models/user_spec.rb]: This issue has been addressed well. Nice job! :+1: :+1:

File: spec/models/post_spec.rb (hunk 1 above)
Comment (henatan99) [reply spec/models/post_spec.rb]: This issue has been addressed well. Nice job! :clap: :clap:

File: README.md (hunk 3 above)
Comment (henatan99) [reply README.md:30]: - [ ] This issue has not been well addressed. I still am getting issues with the seed data. Kindly fix this or may be remove the seed data and you are good to go. :pray: 

Screenshot from 2021-12-10 11-13-21

Screenshot from 2021-12-10 11-21-08

**Review by henatan99 — CHANGES_REQUESTED** [review]
Hi @Kalunge , 

Great work on making the changes requested by a previous reviewer 👏 
You've done well implementing some of the requested changes, but there are still some which aren't addressed yet.


 ## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._


Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**


------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Review by SajjadAhmad14 — APPROVED** [review]
# Status: Approved :heavy_check_mark: :heavy_check_mark: :heavy_check_mark:

### Hi Muthomi, 
## Congratulations 🎉 Thank you for implementing changes suggested previously 🙏

## Happy Coding 👏👏👏
#### Please leave any questions or comments in the PR if you have any doubts or unclarity!
As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s).

**Pull request merged**